src/xrpld/ledger/detail/View.cpp

```diff
@@ -2692,6 +2705,163 @@ sharesToAssetsWithdraw(
+            (*sle)[sfEscrowedAmount] += pay;
+        else
+            sle->setFieldU64(sfEscrowedAmount, pay);
+
+        view.update(sle);
+    }
+
+    // 1. Increase the Issuance EscrowedAmount
+    // 2. DO NOT change the Issuance OutstandingAmount
+    {
+        uint64_t const issuanceEscrowed =
+            (*sleIssuance)[~sfEscrowedAmount].value_or(0);
+        auto const pay = amount.mpt().value();
+
+        // Overflow check for addition
+        if (!canAdd(
+                STAmount(mptIssue, issuanceEscrowed), STAmount(mptIssue, pay)))
+            return tecINTERNAL;
+
+        if (sleIssuance->isFieldPresent(sfEscrowedAmount))
+            (*sleIssuance)[sfEscrowedAmount] += pay;
+        else
+            sleIssuance->setFieldU64(sfEscrowedAmount, pay);
+
+        view.update(sleIssuance);
+    }
+    return tesSUCCESS;
+}
+
+TER
+rippleUnlockEscrowMPT(
+    ApplyView& view,
+    AccountID const& sender,
+    AccountID const& receiver,
+    STAmount const& amount,
+    beast::Journal j)
+{
+    auto const issuer = amount.getIssuer();
+    auto const mptIssue = amount.get<MPTIssue>();
+    auto const mptID = keylet::mptIssuance(mptIssue.getMptID());
+    auto sleIssuance = view.peek(mptID);
+    if (!sleIssuance)
+        return tecOBJECT_NOT_FOUND;
+
+    // Decrease the Issuance EscrowedAmount
+    {
+        if (!sleIssuance->isFieldPresent(sfEscrowedAmount))
+            return tecINTERNAL;
+
+        auto const escrowed = sleIssuance->getFieldU64(sfEscrowedAmount);
+        auto const redeem = amount.mpt().value();
+
+        // Underflow check for subtraction
+        if (!canSubtract(
+                STAmount(mptIssue, escrowed), STAmount(mptIssue, redeem)))
+            return tecINTERNAL;
+
+        sleIssuance->setFieldU64(sfEscrowedAmount, escrowed - redeem);
+        view.update(sleIssuance);
+    }
+
+    if (issuer != receiver)
+    {
+        // Increase the MPT Holder MPTAmount
+        auto const mptokenID = keylet::mptoken(mptID.key, receiver);
+        auto sle = view.peek(mptokenID);
+        if (!sle)
+            return tecOBJECT_NOT_FOUND;
+
+        auto current = sle->getFieldU64(sfMPTAmount);
+        auto delta = amount.mpt().value();
+
+        // Overflow check for addition
+        if (!canAdd(STAmount(mptIssue, current), STAmount(mptIssue, delta)))
+            return tecINTERNAL;
+
+        (*sle)[sfMPTAmount] += delta;
+        view.update(sle);
+    }
+    else
+    {
+        // Decrease the Issuance OutstandingAmount
+        auto const outstanding = sleIssuance->getFieldU64(sfOutstandingAmount);
+        auto const redeem = amount.mpt().value();
+
+        // Underflow check for subtraction
+        if (!canSubtract(
+                STAmount(mptIssue, outstanding), STAmount(mptIssue, redeem)))
+            return tecINTERNAL;
+
+        sleIssuance->setFieldU64(sfOutstandingAmount, outstanding - redeem);
+        view.update(sleIssuance);
+    }
+
+    if (issuer != sender)
```

Comment:
nit: should return tecINTERNAL if issuer == sender

Reply:
fixed: https://github.com/XRPLF/rippled/pull/5185/commits/8774c8a58550d92ab254041f39a14b884eeacd90